File: Modules/RenameOrRenumberFiles.m
function handles = RenameOrRenumberFiles(handles)

% Help for the Rename Or Renumber Files module:
% Category: File Processing
%
% SHORT DESCRIPTION:
% Renames or renumbers files on the hard drive.
% *************************************************************************
%
% This file renaming utility adjusts text within image file names. 
% Be very careful with this module because its purpose is to rename (=
% overwrite) files!! You will have the opportunity to confirm the name
% change for the first cycle only. The folder containing the files must not
% contain subfolders or the subfolders and their contents will also be
% renamed. It is worth doing a practice run with copies of images first.
%
%
% Settings:
%
% * How many characters to retain at the beginning and end of each
% filename? These are the characters that will remain unaltered and note
% that all other characters in between will be removed.
% * The user may choose to add text or numbers between
% the characters that are to be retained.
%
%
% Examples:
% 
% Renumber:
% DrosDAPI_1.tif    -> DrosDAPI_001.tif
% DrosDAPI_10.tif   -> DrosDAPI_010.tif
% DrosDAPI_100.tif  -> DrosDAPI_100.tif
% (to accomplish this, retain 4 characters at the end, retain 9 characters
% at the beginning, and use 3 numerical digits between).
%
% Renumbering is especially useful when numbers within image filenames do
% not have a minimum number of digits and thus appear out of order when
% listed in some Unix/Mac OSX systems. For example, on some systems, files
% would appear like this and be measured out of expected sequence by
% CellProfiler:
% DrosDAPI_1.tif
% DrosDAPI_10.tif
% DrosDAPI_2.tif
% DrosDAPI_3.tif
% DrosDAPI_4.tif
% ...
%
% Rename:
% 1DrosophilaDAPI_1.tif    -> 1DrosDP_1.tif
% 2DrosophilaDAPI_10.tif   -> 2DrosDP_10.tif
% 3DrosophilaDAPI_100.tif  -> 3DrosDP_100.tif
% (to accomplish this, retain 4 characters at the end, retain 5 characters
% at the beginning, enter "DP" as text to place between, and leave
% numerical digits as is).

% CellProfiler is distributed under the GNU General Public License.
% See the accompanying file LICENSE for details.
%
% Developed by the Whitehead Institute for Biomedical Research.
% Copyright 2003,2004,2005.
%
% Please see the AUTHORS file for credits.
%
% Website: http://www.cellprofiler.org
%
% $Revision$

%%%%%%%%%%%%%%%%%
%%% VARIABLES %%%
%%%%%%%%%%%%%%%%%
drawnow


[CurrentModule, CurrentModuleNum, ModuleName] = CPwhichmodule(handles);

% 
%
% Klmadden 2009_03_20 I think this warning should be moved to a header
% position in the variable settings:
% This module allows you to rename (overwrite) your files. Please see the help for this module for warnings.
% So then the first variable should just say:
% What did you call the images you want to rename/renumber?
% 
% See additional PyCP settings below...


%textVAR01 = What did you call the images you want to rename/renumber? Be very careful since you will be renaming (= overwriting) your files!! See the help for this module for other warnings.
%infotypeVAR01 = imagegroup
ImageName = char(handles.Settings.VariableValues{CurrentModuleNum,1});
%inputtypeVAR01 = popupmenu

%textVAR02 = How many characters at the beginning of the file name do you want to retain unaltered?
%defaultVAR02 = 6
NumberCharactersPrefix = str2double(char(handles.Settings.VariableValues{CurrentModuleNum,2}));

%textVAR03 = How many characters at the end do you want to retain unaltered, including the .file extension if present?
%defaultVAR03 = 8
NumberCharactersSuffix = str2double(char(handles.Settings.VariableValues{CurrentModuleNum,3}));

% NEW PyCP setting: What do you want to do with the remaining characters (those not retained above)?
% Options: Delete them all, or Extract the numerical characters
%
% If 'extract numerical characters', setting #5 below should appear.

% NEW PyCP: Should we have the following be "Do you want to add text
% between the beginning and ending portions of filename?" and if yes then
% ask what text?  And, this variable (4) should be placed after (5) - we
% want to address the numerical characters first since we just asked about
% them in the New setting above. I think this means the code should be
% rearranged a bit because I bet right now the text is automatically placed first and the
% numbers second (which is the most useful scenario, I think). So, as I've
% rewritten the settings, rearranging 4 and 5, it appears that numbers will
% come first and text second - should we
% also offer the user the choice of putting text first and numbers second?
%
%textVAR04 = Enter any text you want to place between the beginning and ending portions of the filename. Type  "Do not use" to leave as is.
%defaultVAR04 = Do not use
TextToAdd = char(handles.Settings.VariableValues{CurrentModuleNum,4});

%New PyCP: The text "Type Do not use to leave as is." should no longer be
%necessary (because #5 only appears if 'extract numerical chars" was
%chosen).
%
%textVAR05 = How many numerical digits would you like to use between the beginning and ending portions of the filename, for renumbering purposes? Type "Do not use" to leave as is.
%defaultVAR05 = Do not use
NumberDigits = char(handles.Settings.VariableValues{CurrentModuleNum,5});

%%%VariableRevisionNumber = 1

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% PRELIMINARY CALCULATIONS & FILE HANDLING %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
drawnow

Pathname = handles.Current.DefaultImageDirectory;
%%% Retrieves all the image file names and the number of
%%% images per set so they can be used by the module.
fieldname=['FileList',ImageName];
FileNames = handles.Pipeline.(fieldname);
if ~ischar(TextToAdd)
    try TextToAdd = num2str(TextToAdd);
    catch error(['Image processing was canceled in the ', ModuleName, ' module because the text you tried to add could not be converted into text for some reason.'])
    end
end

AmtDigits = str2double(NumberDigits);

for n = 1:length(FileNames)
    OldFilename = char(FileNames(n));
    Prefix = OldFilename(1:NumberCharactersPrefix);
    Suffix = OldFilename((end-NumberCharactersSuffix+1):end);

    %Renumbering Stage
    if ~strcmp(NumberDigits,'Do not use')
        OldNumber = OldFilename(NumberCharactersPrefix+1:end-NumberCharactersSuffix);
        NumberOfZerosToAdd = AmtDigits - length(OldNumber);
        if NumberOfZerosToAdd < 0
            OldNumber = OldNumber(end-AmtDigits+1:end);
            NumberOfZerosToAdd=0;
        end
        ZerosToAdd = num2str(zeros(NumberOfZerosToAdd,1))';
        NewText = [ZerosToAdd,OldNumber];
    else
        NewText = '';
    end

    %Renaming Stage
    if strcmp(TextToAdd,'Do not use')
        NewFilename = [Prefix,NewText,Suffix];
    else
        NewFilename = [Prefix,TextToAdd,NewText,Suffix];
    end

    if n == 1
        DialogText = ['Confirm the file name change. For example, the first file''s name will change from ', OldFilename, ' to ', NewFilename, '.  The remaining files will be converted without asking for confirmation.'];
        Answer = CPquestdlg(DialogText, 'Confirm file name change','OK','Cancel','Cancel');
        if strcmp(Answer, 'Cancel')
            error(['Image processing was canceled in the ', ModuleName, ' module at your request.'])
        end
    end
    if ~strcmp(OldFilename,NewFilename)
        movefile(fullfile(Pathname,OldFilename),fullfile(Pathname,NewFilename))
    end
    drawnow
end

%%% This line will "cancel" processing after the first time through this
%%% module.  All the files are renumbered the first time through. Without
%%% the following cancel line, the module will run X times, where X is the
%%% number of files in the current directory.
set(handles.timertexthandle,'string','Cancel')

%%%%%%%%%%%%%%%%%%%%%
%%% FIGURE WINDOW %%%
%%%%%%%%%%%%%%%%%%%%%
drawnow

%%% The figure window display is unnecessary for this module, so it is
%%% closed during the starting image cycle.
CPclosefigure(handles,CurrentModule)

File: pyCellProfiler/cellprofiler/modules/renameorrenumberfiles.py
'''<b>RenameOrRenumberFiles</b> - Renames or renumbers files on the hard drive.
<hr>

This file renaming utility adjusts text within image file names. 
<i style='color:red'>Be very careful with this module because its purpose is 
to rename (= overwrite) files!!</i> You will have the opportunity to confirm 
the name change for the first cycle only. The folder containing the files must 
not contain subfolders or the subfolders and their contents will also be
renamed. The module will not rename the file in test mode. You should use
test mode to ensure that the module settings are correct.

You can use this module to standardize the number of characters in your
file names and to remove unwanted characters from your file names. This is
especially useful if you want file names that have numbers in them to appear
in numerical order when processed by LoadImages.<br>
<br>
Examples:<br>
<br>
Renumber:<br>
DrosDAPI_1.tif    -> DrosDAPI_001.tif<br>
DrosDAPI_10.tif   -> DrosDAPI_010.tif<br>
DrosDAPI_100.tif  -> DrosDAPI_100.tif<br>
(to accomplish this, retain 4 characters at the end, retain 9 characters
at the beginning, and use 3 numerical digits between).<br>
<br>
Renumbering is especially useful when numbers within image filenames do
not have a minimum number of digits and thus appear out of order when
listed in some Unix/Mac OSX systems. For example, on some systems, files
would appear like this and be measured out of expected sequence by
CellProfiler:<br>
DrosDAPI_1.tif<br>
DrosDAPI_10.tif<br>
DrosDAPI_2.tif<br>
DrosDAPI_3.tif<br>
DrosDAPI_4.tif<br>
...<br>
<br>
Rename:<br>
1DrosophilaDAPI_1.tif    -> 1DrosDP_1.tif<br>
2DrosophilaDAPI_10.tif   -> 2DrosDP_10.tif<br>
3DrosophilaDAPI_100.tif  -> 3DrosDP_100.tif<br>
(to accomplish this, retain 4 characters at the end, retain 5 characters
at the beginning, enter "DP" as text to place between, and leave
numerical digits as is).
'''

__version__="$Revision$"

import os

import cellprofiler.cpmodule as cpm
import cellprofiler.settings as cps

A_RENUMBER = "Renumber"
A_DELETE = "Delete"

class RenameOrRenumberFiles(cpm.CPModule):
    
    module_name = "RenameOrRenumberFiles"
    category = "File Processing"
    variable_revision_number = 1
    
    def create_settings(self):
        '''Create the settings for the module's UI'''
        self.warning = cps.Divider(
            "This module allows you to rename (overwrite) your files. Please "
            "see the help for this module for warnings.")
        self.image_name = cps.FileImageNameSubscriber(
            'Image name:','None',
            doc="""This is the name of the image associated with the file
            you want to rename. It should be an image loaded by 
            <b>LoadImages</b>, <b>LoadData</b> or <b>LoadSingleImage</b>.
            Be very careful because you will be renaming this file!""")
        self.number_characters_prefix = cps.Integer(
            "Number of characters to retain at start of file name:", 6,
            minval=0,
            doc="""This is the number of characters at the start of the old
            file name that will be copied over to the new file name. For
            instance, if this setting is "6" and the file name is 
            "Image-734.tif", the output file name will also start with 
            "Image-".""")
        self.number_characters_suffix = cps.Integer(
            "Number of characters to retain at the end of file name:", 4,
            minval=0,
            doc="""This is the number of characters at the end of the old
            file name that will be copied over to the new file name. For
            instance, if this setting is "4" and the file name is
            "Image-734.tif", the output file name will also end with ".tif".""")
        self.action = cps.Choice(
            "What do you want to do with the remaining characters?",
            [A_RENUMBER, A_DELETE],
            doc ="""You can either treat the characters between the start and
            end as numbers or you can delete them. If you treat them as numbers,
            you will be given the opportunity to pad the numbers with zeros
            so that all of your file names will have a uniform length. For
            instance, if you were to renumber the highlighted portion of the
            file, "Image-<u>734</u>.tif", using four digits, the result would
            be, "Image-0734.tif".""")
        self.number_digits = cps.Integer(
            "How many numerical digits would you like to use?",
            4, minval=0,
            doc="""Use this setting to pad numbers with zeros so that they
            all have a uniform number of characters. For instance, padding
            with four digits has the following result:<br>
            <code><table>
            <tr><th>Original</th><th>Padded</th></tr>
            <tr><td>1</td><td>0001</td></tr>
            <tr><td>10</td><td>0010</td></tr>
            <tr><td>100</td><td>0100</td></tr>
            <tr><td>1000</td><td>1000</td></tr>
            </table></code>""")
        self.wants_text = cps.Binary(
            "Do you want to add text to the file name?", False,
            doc="""You can check this setting if you want to add text
            to the file name. If you chose "Renumber" above,
            <b>RenameOrRenumberFiles</b> will add the text after your number.
            If you chose "Delete", <b>RenameOrRenumberFiles</b> will replace
            the deleted text with the text you enter here.""")
        self.text_to_add = cps.Text(
            "Replacement text","",
            doc="""This is the text that appears either after your number or
            instead of the deleted text.""")
            
    def settings(self):
        '''Return settings in the order that they should appear in pipeline'''
        return [ self.image_name, self.number_characters_prefix,
                 self.number_characters_suffix, self.action, 
                 self.number_digits, self.wants_text, self.text_to_add]
    
    def visible_settings(self):
        '''Return the settings to display in the GUI'''
        result = [ self.warning, self.image_name, self.number_characters_prefix,
                   self.number_characters_suffix, self.action]
        if self.action == A_RENUMBER:
            result += [self.number_digits]
        result += [self.wants_text]
        if self.wants_text:
            result += [self.text_to_add]
        return result
    
    def is_interactive(self):
        '''Tell everyone that the GUI is not an interactive one'''
        return False
    
    def run(self, workspace):
        '''Run on an image set'''
        image_name = self.image_name.value
        m = workspace.measurements
        #
        # Get the path and file names from the measurements
        #
        path = m.get_current_image_measurement('PathName_%s' % image_name)
        file_name = m.get_current_image_measurement('FileName_%s' % image_name)
        #
        # Pull out the prefix, middle and suffix from the file name
        #
        prefix = file_name[:self.number_characters_prefix.value]
        if self.number_characters_suffix.value == 0:
            suffix = ""
            middle = file_name[self.number_characters_prefix.value:]
        else:
            suffix = file_name[-self.number_characters_suffix.value:]
            middle = file_name[self.number_characters_prefix.value:
                               -self.number_characters_suffix.value]
        #
        # Possibly apply the renumbering rule
        #
        if self.action == A_RENUMBER:
            if not middle.isdigit():
                raise ValueError(
                    ('The middle of the filename, "%s", is "%s".\n'
                     "It has non-numeric characters and can't be "
                     "converted to a number") %
                    ( file_name, middle ))
            format = '%0'+str(self.number_digits.value)+'d'
            middle = format % int(middle)
        elif self.action == A_DELETE:
            middle = ""
        else:
            raise NotImplementedError("Unknown action: %s" % self.action.value)
        #
        # Possibly apply the added text
        #
        if self.wants_text:
            middle += self.text_to_add.value
        new_file_name = prefix + middle + suffix
        if workspace.frame is not None:
            workspace.display_data.old_file_name = file_name
            workspace.display_data.new_file_name = new_file_name
        
        if workspace.pipeline.test_mode:
            return
        #
        # Perform the actual renaming
        #
        os.rename(os.path.join(path, file_name),
                  os.path.join(path, new_file_name))
        
    def display(self, workspace):
        '''Display the pathname conversion'''
        statistics = [('Old file name','New file name'),
                      (workspace.display_data.old_file_name,
                       workspace.display_data.new_file_name)]
        figure = workspace.create_or_find_figure(subplots=(1,1))
        figure.subplot_table(0,0,statistics, ratio = (.5, .5))

    def upgrade_settings(self, setting_values, variable_revision_number,
                         module_name, from_matlab):
        '''Upgrade settings from previous pipeline versions
        
        setting_values - string values for each of the settings
        variable_revision_number - the revision number of the module at the
                                   time of saving
        module_name - the name of the module that saved the settings
        from_matlab - true if pipeline was saved by CP 1.0
        '''
        if from_matlab and variable_revision_number == 1:
            image_name, number_characters_prefix, number_characters_suffix,\
            text_to_add, number_digits = setting_values
            
            wants_text = text_to_add.lower() != cps.DO_NOT_USE.lower()
            wants_text = cps.YES if wants_text else cps.NO
            
            if number_digits.lower() == cps.DO_NOT_USE.lower():
                number_digits = 4
                action = A_DELETE
            else:
                action = A_RENUMBER
            
            setting_values = [image_name, number_characters_prefix, 
                              number_characters_suffix, action,
                              number_digits, wants_text, text_to_add]
            variable_revision_number = 1
            from_matlab = False
        return setting_values, variable_revision_number, from_matlab